### PR TITLE
Protocol: make `supported_sync_modes` a required not empty list on `AirbyteStream`

### LINKED ISSUE #17563: protocol: `supported_sync_modes` is not required on `AirbyteStream`
## Tell us about the problem you're trying to solve
Our platform and source acceptance test expect the `AirbyteStream` instances to have a `supported_sync_modes` not null attribute but this is not enforced at the protocol.

## Describe the solution you’d like
I want the `supported_sync_modes` attribute to be a required field on our protocol's `AirbyteStream` model.

## Describe the alternative you’ve considered or used
Enforce this rule with the source-acceptance tests, but SAT does not currently cover all our connectors.

## Additional context
[Add any other context or screenshots about the feature request here.
](https://github.com/airbytehq/airbyte/issues/14547)

## Are you willing to submit a PR?
https://github.com/airbytehq/airbyte/pull/15591

### PR DESCRIPTION
**Pull request by alafanechere**
Closes https://github.com/airbytehq/airbyte/issues/17563

## What
#14547 raised the awareness that the `supported_sync_modes` field in `AirbyteStream` is not a required field and it can be empty. This can lead to unexpected UI behavior as described in the issue. 
We added a  test to SAT (https://github.com/airbytehq/airbyte/pull/14800) to ensure the `supported_sync_mode` is not an empty list, but this is something that can be enforced at the protocol level.

## How
* Change [airbyte_protocol.yaml](https://github.com/airbytehq/airbyte/blob/master/airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml#L221) to make the `supported_sync_mode` field of `AirbyteStream` a **required not empty array**.
* Update SAT and CDK accordingly.
* Make changes on impacted source connectors (`source-google-sheet` and `source-azure-table`).

## Suggested review order

1. [airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml](https://github.com/airbytehq/airbyte/pull/15591/files#diff-23f6bc12d2cf0cbb40771b96abb371d79d04b0eeccfe92d276f314325600b029)
2. [airbyte-cdk/python/airbyte_cdk/models/airbyte_protocol.py](https://github.com/airbytehq/airbyte/pull/15591/files#diff-a22224a3797701e1a0a0e8e53eed63c83ebae312c2c174dbbf5a4cbbbca0739c)

## 🚨 User Impact 🚨
I'm aware that protocol updates are critical. I did not file a tech spec as I'm under the impression this modification is a fix and not a protocol evolution. @cgardens mentioned in this [comment](https://github.com/airbytehq/airbyte/blob/master/airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml#L227) that this change can happen *once sources are migrated*. I'd love to have more context about this comment to evaluate the potential impact of this change.

### CODE REVIEW

File: airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml [hunk 1]
```diff
@@ -224,8 +224,7 @@ definitions:
     required:
       - name
       - json_schema
-      # todo (cgardens) - make required once sources are migrated
-      # - supported_sync_modes
+      - supported_sync_modes
```
Comment (evantahler) [inline airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml:227]: I guess we've migrated!

Comment (evantahler) [inline airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml:227]: I guess we've migrated!  Can you confirm that there are no old source that don't provide this property?

Comment (alafanechere) [reply airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml:227]: I'm not 💯 of the Singer compatibility...

Comment (alafanechere) [reply airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml:227]: > Can you confirm that there are no old source that don't provide this property?
We can double check that existing GA/beta connectors SAT is passing - a test to check the presence of this property was shipped a week ago #14800

File: airbyte-cdk/python/unit_tests/singer/test_singer_source.py [hunk 2]
```diff
@@ -95,7 +95,7 @@ def test_singer_discover_metadata(mock_read_catalog):
     _user_stream = airbyte_catalog.streams[0]
     _roles_stream = airbyte_catalog.streams[1]
 
-    assert _user_stream.supported_sync_modes is None
+    # assert _user_stream.supported_sync_modes is None
```
Comment (girarda) [inline airbyte-cdk/python/unit_tests/singer/test_singer_source.py:98]: should this assert the right modes are supported?

**Review by girarda — APPROVED** [review]
:shipit:

File: airbyte-cdk/python/unit_tests/singer/test_singer_source.py (hunk 2 above)
Comment (alafanechere) [reply airbyte-cdk/python/unit_tests/singer/test_singer_source.py:98]: I'm not sure why this assertion was declared. There is probably an edge case with singer and supported sync modes that I'm unaware of.

**Review by sherifnada — APPROVED** [review]

**Comment by alafanechere** [comment]
This PR will require:
1. Publishing changed made an `source-google-sheet` and `source-azure-table`
2. Publishing SAT
3. Publishing CDK

File: airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml (hunk 1 above)
Comment (evantahler) [reply airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml:227]: I'm going to add @jdpgrailsdev as well to confirm, but I think this is 👍

**Review by evantahler — APPROVED** [review]

**Review by jdpgrailsdev — APPROVED** [review]
:shipit:

**Comment by alafanechere** [comment]
Platform build did not pass because of acceptance tests failures:
* They rely on an old version of the source-e2e-test connector, which is not respecting my protocol change
* I published the new version of source-e2e-test  and updated the hardcoded version [here](https://github.com/airbytehq/airbyte/blob/de0da4bc8388054fbaba0c264a7dd32f13325c0a/airbyte-test-utils/src/main/java/io/airbyte/test/utils/AirbyteAcceptanceTestHarness.java#L129)
* I updated the config passed to the connector [here](https://github.com/airbytehq/airbyte/blob/de0da4bc8388054fbaba0c264a7dd32f13325c0a/airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AdvancedAcceptanceTests.java#L259)

Hopefully they'll pass now 🙏

**Comment by alafanechere** [comment]
/publish connector=connectors/source-azure-table,connectors/source-google-sheet

**Comment by alafanechere** [comment]
/publish connector=bases/source-acceptance-test auto-bump-version=false

**Comment by grubberr** [comment]
@alafanechere 
https://github.com/airbytehq/airbyte/actions/runs/3279207635/jobs/5398502934#step:11:1089

**Pull request merged**